enderio-machines/src/main/java/com/enderio/machines/common/blocks/wireless_charger/WirelessChargerBlockEntity.java

```diff
@@ -0,0 +1,230 @@
+        super.neighborChanged(neighborBlock, neighborPos);
+        if (level != null && !level.isClientSide() && getBlockPos().above().equals(neighborPos)) {
+            int maxRange = calculateMaxRange();
+            if (getRangeExtension() > 0 || actionRange.range() > maxRange) {
+                // Antenna placed or removed from the top so update range
+                setActionRange(new ActionRange(maxRange, actionRange.isVisible(), maxRange));
+            }
+        }
+    }
+
+    @Override
+    public void serverTick() {
+        super.serverTick();
+        if (isActive()) {
+            getEnergyStorage().consumeEnergy(energyUpkeep.get());
+            chargeItem();
+        }
+    }
+
+    @Override
+    public void clientTick() {
+        if (level instanceof ClientLevel clientLevel) {
+            getActionRange().addClientParticle(clientLevel, getBlockPos(),
+                    MachinesConfig.CLIENT.BLOCKS.WIRELESS_CHARGER_RANGE_COLOR.get());
+        }
+        super.clientTick();
+    }
+
+    @Override
+    public void onLoad() {
+        super.onLoad();
+        if (level != null && !level.isClientSide() && actionRange.range() > calculateMaxRange()) {
+            int maxRange = calculateMaxRange();
+            setActionRange(new ActionRange(maxRange, actionRange.isVisible(), maxRange));
+        }
+        updateBounds();
+    }
+
+    @Override
+    protected void saveAdditionalSynced(CompoundTag tag, HolderLookup.Provider registries) {
+        super.saveAdditionalSynced(tag, registries);
+        tag.put(MachineNBTKeys.ACTION_RANGE, actionRange.save(registries));
+    }
+
+    @Override
+    protected void loadAdditional(CompoundTag tag, HolderLookup.Provider registries) {
+        super.loadAdditional(tag, registries);
+        if (tag.contains(MachineNBTKeys.ACTION_RANGE)) {
+            actionRange = ActionRange.parse(registries, Objects.requireNonNull(tag.get(MachineNBTKeys.ACTION_RANGE)));
+        } else {
+            int maxRange = calculateMaxRange();
+            actionRange = new ActionRange(maxRange, false, maxRange);
+        }
+    }
+
+    @Override
+    protected void applyImplicitComponents(DataComponentInput components) {
+        super.applyImplicitComponents(components);
+        var actionRange = components.get(MachineDataComponents.ACTION_RANGE);
+        if (actionRange != null) {
+            this.actionRange = actionRange;
+        }
+    }
+
+    @Override
+    protected void collectImplicitComponents(DataComponentMap.Builder components) {
+        super.collectImplicitComponents(components);
+        components.set(MachineDataComponents.ACTION_RANGE, actionRange);
+    }
+
+    @Override
+    public void removeComponentsFromTag(CompoundTag tag) {
+        super.removeComponentsFromTag(tag);
+        tag.remove(MachineNBTKeys.ACTION_RANGE);
+    }
+
+    @Override
+    public boolean isActive() {
+        return hasEnergy() && canAct();
+    }
+
+    private int calculateMaxRange() {
```

Comment:
Instead of storing maxRange in the ActionRange, just cache the max range whenever the neighbours change.